src/eduid/webapp/common/session/logindata.py

```diff
@@ -35,69 +40,169 @@ class ExternalMfaData(BaseModel):
+
+
+TLoginContextSubclass = TypeVar('TLoginContextSubclass', bound='LoginContext')
+
+
+@dataclass
+class LoginContextSAML(LoginContext):
+
+    _saml_req: Optional['IdP_SAMLRequest'] = field(default=None, init=False, repr=False)
 
     @property
     def SAMLRequest(self) -> str:
-        if not isinstance(self.saml_data.request, str):
+        pending = self.pending_request
+        if not isinstance(pending, IdP_SAMLPendingRequest):
+            raise ValueError('Pending request not initialised (or not a SAML request)')
+        if not isinstance(pending.request, str):
             raise ValueError('saml_data.request not initialised')
-        return self.saml_data.request
+        return pending.request
 
     @property
     def RelayState(self) -> str:
-        if not self.saml_data.relay_state:
-            return ''
-        return self.saml_data.relay_state
+        pending = self.pending_request
+        if not isinstance(pending, IdP_SAMLPendingRequest):
+            raise ValueError('Pending request not initialised (or not a SAML request)')
+        return pending.relay_state or ''
 
     @property
     def binding(self) -> str:
-        if not isinstance(self.saml_data.binding, str):
+        pending = self.pending_request
+        if not isinstance(pending, IdP_SAMLPendingRequest):
+            raise ValueError('Pending request not initialised (or not a SAML request)')
+        if not isinstance(pending.binding, str):
             raise ValueError('saml_data.binding not initialised')
-        return self.saml_data.binding
+        return pending.binding
 
     @property
     def query_string(self) -> str:
         qs = {'SAMLRequest': self.SAMLRequest, 'RelayState': self.RelayState}
         return urlencode(qs)
 
     @property
-    def saml_req(self) -> 'IdP_SAMLRequest':
+    def saml_req(self) -> IdP_SAMLRequest:
         if self._saml_req is None:
-            raise ValueError('SSOLoginData.saml_req accessed uninitialised')
+            # avoid circular import
+            from eduid.webapp.idp.app import current_idp_app as current_app
+
+            self._saml_req = IdP_SAMLRequest(
+                self.SAMLRequest, self.binding, current_app.IDP, debug=current_app.conf.debug
+            )
         return self._saml_req
 
-    @saml_req.setter
-    def saml_req(self, value: Optional['IdP_SAMLRequest']):
-        self._saml_req = value
+    @property
+    def request_id(self) -> Optional[str]:
+        return self.saml_req.request_id
+
+    @property
+    def authn_contexts(self) -> List[str]:
+        return self.saml_req.get_requested_authn_contexts()
+
+    @property
+    def reauthn_required(self) -> bool:
+        return self.saml_req.force_authn
+
+    @property
+    def other_device_state_id(self) -> Optional[OtherDeviceId]:
+        # On device #1, the pending_request has a pointer to the other-device-state
+        _pending = self.pending_request
```

Comment:
added comment explaining the use of _pending:
        # Use temporary variable to avoid pycharm warning
        #   Unresolved attribute reference 'other_device_state_id' for class 'IdP_PendingRequest'